spyderlib/widgets/file_switcher.py

```diff
@@ -0,0 +1,415 @@
+            recurse_level({idx: toks[k:] for idx, toks in group.iteritems()})
+
+    recurse_level(dict(enumerate(path_list)))
+    return [path[:-1] for path in new_path_list]  # trim final sep
+
+
+class HTMLDelegate(QStyledItemDelegate):
+    """With this delegate, a QListWidgetItem can render HTML.
+
+    Taken from http://stackoverflow.com/a/5443112/2399799
+    """
+    def paint(self, painter, option, index):
+        options = QStyleOptionViewItemV4(option)
+        self.initStyleOption(options, index)
+
+        style = QApplication.style() if options.widget is None\
+            else options.widget.style()
+
+        doc = QTextDocument()
+        doc.setHtml(options.text)
+
+        options.text = ""
+        style.drawControl(QStyle.CE_ItemViewItem, options, painter)
+
+        ctx = QAbstractTextDocumentLayout.PaintContext()
+
+        textRect = style.subElementRect(QStyle.SE_ItemViewItemText, options)
+        painter.save()
+        painter.translate(textRect.topLeft())
+        painter.setClipRect(textRect.translated(-textRect.topLeft()))
+        doc.documentLayout().draw(painter, ctx)
+
+        painter.restore()
+
+    def sizeHint(self, option, index):
+        options = QStyleOptionViewItemV4(option)
+        self.initStyleOption(options, index)
+
+        doc = QTextDocument()
+        doc.setHtml(options.text)
+        doc.setTextWidth(options.rect.width())
+        return QSize(doc.idealWidth(), doc.size().height())
+
+
+class UpDownFilter(QObject):
+    """Use with ``installEventFilter`` to get up/down arrow key press signal.
+    """
+    up_down = Signal(int)
+
+    def eventFilter(self, src, e):
+        if isinstance(e, QKeyEvent) and e.type() == QEvent.KeyPress:
+            if e.key() == Qt.Key_Up:
+                self.up_down.emit(-1)
+            elif e.key() == Qt.Key_Down:
+                self.up_down.emit(+1)
+        return super(UpDownFilter, self).eventFilter(src, e)
+
+
+class HelperToolButton(QToolButton):
+    """ """
+    def __init__(self):
+        QToolButton.__init__(self)
+
+    def setToolTip(self, text):
+        """ """
+        self._tip_text = text
+
+    def toolTip(self):
+        """ """
+        return self._tip_text
+
+    def mousePressEvent(self, event):
+        """ """
+        QToolTip.hideText()
+
+    def mouseReleaseEvent(self, event):
+        """ """
+        QToolTip.showText(self.mapToGlobal(QPoint(0, 0)), self._tip_text)
+
+
+class FileSwitcher(QDialog):
+    close_file = Signal(int)
+    edit_file = Signal(int)
+    edit_line = Signal(int)
+
+    def __init__(self, parent, tabs, tab_data):
+        QDialog.__init__(self, parent)
+
+        # Destroying the C++ object right after closing the dialog box,
+        # otherwise it may be garbage-collected in another QThread
+        # (e.g. the editor's analysis thread in Spyder), thus leading to
+        # a segmentation fault on UNIX or an application crash on Windows
+        self.setAttribute(Qt.WA_DeleteOnClose)
```

Comment:
That is true, but it doesn't seem to be much of an issue.  I inherited this version from the original, and I never bothered changing it.